File: ios/Classes/lib/awesome_notifications_core/managers/CancellationManager.swift
//
//  CancellationManager.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 06/12/21.
//

import Foundation

public class CancellationManager {
    
    public static let TAG: String = "CancellationManager"
    
    // ************** SINGLETON PATTERN ***********************
    
    public static let shared: CancellationManager = CancellationManager()
    private init(){}
    
    // ********************************************************
        
    public func dismissNotification(byId id:Int) -> Bool {
        return StatusBarManager.dismissNotification(id: id)
    }
    
    public func dismissNotifications(byChannelKey channelKey: String) -> Bool {
        return StatusBarManager.dismissNotificationsByChannelKey(channelKey: channelKey)
    }
    
    public func dismissNotifications(byGroupKey groupKey: String) -> Bool {
        return StatusBarManager.dismissNotificationsByGroupKey(groupKey: groupKey)
    }
    
    public func dismissAllNotifications() -> Bool {
        return StatusBarManager.dismissAllNotifications()
    }
    
    public func cancelSchedule(byId id:Int) -> Bool {
        return
            ScheduleManager.cancelScheduled(id: id) &&
            _cancelNativeScheduledNotification(id: id)
    }
    
    public func cancelSchedules(byChannelKey channelKey: String) -> Bool {
        return
            _cancelNativeSchedulesByChannelKey(channelKey: channelKey)
    }
    
    public func cancelSchedules(byGroupKey groupKey: String) -> Bool {
        return
            _cancelNativeSchedulesByGroupKey(groupKey: groupKey)
    }
    
    public func cancelAllSchedules() -> Bool {
        return
            ScheduleManager.cancelAllSchedules() &&
            _cancellAllNativeScheduledNotifications()
    }
    
    public func cancelNotification(byId id:Int) -> Bool {
        return
            dismissNotification(id: id) &&
            cancelSchedule(id: id)
    }
    
    public func cancelNotifications(byChannelKey channelKey: String) -> Bool {
        return
            dismissNotificationsByChannelKey(channelKey: channelKey) &&
            cancelSchedulesByChannelKey(channelKey: channelKey)
    }
    
    public func cancelNotifications(byGroupKey groupKey: String) -> Bool {
        return
            dismissNotificationsByGroupKey(groupKey: groupKey) &&
            cancelSchedulesByGroupKey(groupKey: groupKey)
    }

    public func cancelAllNotifications() -> Bool {
        return
            dismissAllNotifications() &&
            cancelAllSchedules()
    }
        
    private func _cancelNativeSchedulesByChannelKey(channelKey: String) -> Bool {
        
        let center = UNUserNotificationCenter.current()
        center.getPendingNotificationRequests(completionHandler: { (notificationRequest) in
            for notification in notificationRequest {
                if channelKey == notification.content.userInfo[Definitions.NOTIFICATION_CHANNEL_KEY] as? String {
                    if notification.content.userInfo[Definitions.NOTIFICATION_ID] != nil {
                        
                        if let id:String = notification.content.userInfo[Definitions.NOTIFICATION_ID] as? String {
                            center.removePendingNotificationRequests(withIdentifiers: [id])
                        }
                        else {
                            if let id:Int64 = notification.content.userInfo[Definitions.NOTIFICATION_ID] as? Int64 {
                                center.removePendingNotificationRequests(withIdentifiers: [String(id)])
                            }
                        }
                    }
                }
            }
        })
        
        return true
    }
    
    private func _cancelNativeSchedulesByGroupKey(groupKey: String) -> Bool {

        let center = UNUserNotificationCenter.current()
        center.getPendingNotificationRequests(completionHandler: { (notificationRequest) in
            for notification in notificationRequest {
                if groupKey == notification.content.userInfo[Definitions.NOTIFICATION_GROUP_KEY] as? String {
                    if notification.content.userInfo[Definitions.NOTIFICATION_ID] != nil {
                        
                        if let id:String = notification.content.userInfo[Definitions.NOTIFICATION_ID] as? String {
                            center.removePendingNotificationRequests(withIdentifiers: [id])
                        }
                        else {
                            if let id:Int64 = notification.content.userInfo[Definitions.NOTIFICATION_ID] as? Int64 {
                                center.removePendingNotificationRequests(withIdentifiers: [String(id)])
                            }
                        }
                    }
                }
            }
        })
        
        return true
    }
    
    private func _cancelNativeScheduledNotification(id:Int) -> Bool {
        let referenceKey:String = String(id)
            
        let center = UNUserNotificationCenter.current()
        center.removePendingNotificationRequests(withIdentifiers: [referenceKey])
        
        return true
    }
    
    private func _cancelNativeNotification(id:Int) -> Bool {
        let referenceKey:String = String(id)
            
        let center = UNUserNotificationCenter.current()
        center.removeDeliveredNotifications(withIdentifiers: [referenceKey])
        center.removePendingNotificationRequests(withIdentifiers: [referenceKey])
        
        return true
    }
    
    private func _cancellAllNativeScheduledNotifications() -> Bool {
            
        let center = UNUserNotificationCenter.current()
        center.removeAllPendingNotificationRequests()
        
        return true
    }
    
    public func _cancellAllNativeNotifications() -> Bool {
            
        let center = UNUserNotificationCenter.current()
        center.removeAllDeliveredNotifications()
        center.removeAllPendingNotificationRequests()
        
        return true
    }
    
}

File: android/src/main/java/me/carda/awesome_notifications/awesome_notifications_core/managers/CancellationManager.java
package me.carda.awesome_notifications.awesome_notifications_core.managers;

import android.content.Context;

import androidx.annotation.NonNull;
import me.carda.awesome_notifications.awesome_notifications_core.threads.NotificationScheduler;
import me.carda.awesome_notifications.awesome_notifications_core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.awesome_notifications_core.utils.BitmapUtils;
import me.carda.awesome_notifications.awesome_notifications_core.utils.StringUtils;

public class CancellationManager {

    private static final String TAG = "CancellationManager";

    private final StringUtils stringUtils;
    
    // ************** SINGLETON PATTERN ***********************

    protected static CancellationManager instance;

    protected CancellationManager(StringUtils stringUtils){
        this.stringUtils = stringUtils;
    }

    public static CancellationManager getInstance() {
        if (instance == null)
            instance = new CancellationManager(
                    StringUtils.getInstance());
        return instance;
    }

    // ********************************************************

    public boolean dismissNotification(@NonNull final Context context, Integer notificationId) throws AwesomeNotificationsException {

        if (notificationId == null || notificationId < 0)
            throw new AwesomeNotificationsException("Invalid notification id");

        StatusBarManager
                .getInstance(context)
                .dismissNotification(notificationId);

        return true;
    }

    public boolean dismissNotificationsByChannelKey(@NonNull final Context context, @NonNull final String channelKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(channelKey))
            throw new AwesomeNotificationsException("Invalid channel key");

        StatusBarManager
                .getInstance(context)
                .dismissNotificationsByChannelKey(channelKey);

        return true;
    }

    public boolean dismissNotificationsByGroupKey(@NonNull final Context context, @NonNull final String groupKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(groupKey))
            throw new AwesomeNotificationsException("Invalid group key");

        StatusBarManager
                .getInstance(context)
                .dismissNotificationsByGroupKey(groupKey);

        return true;
    }

    public void dismissAllNotifications(@NonNull final Context context) {

        StatusBarManager
                .getInstance(context)
                .dismissAllNotifications();
    }

    public boolean cancelSchedule(@NonNull final Context context, Integer notificationId) throws AwesomeNotificationsException {

        if (notificationId == null || notificationId < 0)
            throw new AwesomeNotificationsException("Invalid notification id");

        NotificationScheduler.cancelScheduleById(context, notificationId);

        return true;
    }

    public boolean cancelSchedulesByChannelKey(@NonNull final Context context, @NonNull final String channelKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(channelKey))
            throw new AwesomeNotificationsException("Invalid channel key");

        NotificationScheduler.cancelSchedulesByChannelKey(context, channelKey);

        return true;
    }

    public boolean cancelSchedulesByGroupKey(@NonNull final Context context, @NonNull final String groupKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(groupKey))
            throw new AwesomeNotificationsException("Invalid group key");

        NotificationScheduler.cancelSchedulesByGroupKey(context, groupKey);

        return true;
    }

    public void cancelAllSchedules(@NonNull final Context context) {
        NotificationScheduler.cancelAllSchedules(context);
    }

    public boolean cancelNotification(@NonNull final Context context, Integer notificationId) throws AwesomeNotificationsException {

        if (notificationId == null || notificationId < 0)
            throw new AwesomeNotificationsException("Invalid notification id");

        cancelSchedule(context, notificationId);
        dismissNotification(context, notificationId);

        return true;
    }

    public boolean cancelNotificationsByChannelKey(@NonNull final Context context, @NonNull final String channelKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(channelKey))
            throw new AwesomeNotificationsException("Invalid channel key");

        dismissNotificationsByChannelKey(context, channelKey);
        cancelSchedulesByChannelKey(context, channelKey);

        return true;
    }

    public boolean cancelNotificationsByGroupKey(@NonNull final Context context, @NonNull final String groupKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(groupKey))
            throw new AwesomeNotificationsException("Invalid group key");

        dismissNotificationsByGroupKey(context, groupKey);
        cancelSchedulesByGroupKey(context, groupKey);

        return true;
    }


    public void cancelAllNotifications(@NonNull final Context context) {

        dismissAllNotifications(context);
        cancelAllSchedules(context);
    }
}
